iwyu.cc

```diff
@@ -2201,24 +2201,55 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // an iwyu warning now, even without knowing the exact overload.
   // In that case, we store the fact we warned, so we won't warn again
   // when the template is instantiated.
-  // TODO(csilvers): to be really correct, we should report *every*
-  // overload that callers couldn't match via ADL.
+  // Otherwise, all directly '#include'd overloads are reported just to
+  // be kept.
   bool VisitUnresolvedLookupExpr(UnresolvedLookupExpr* expr) {
     // No CanIgnoreCurrentASTNode() check here!  It's later in the function.
 
-    // Make sure all overloads are in the same file.
     if (expr->decls_begin() == expr->decls_end()) {
       // This can occur when there are no overloads before the template
       // definition, and a callee may be found only via ADL.
       return true;
     }
     const NamedDecl* first_decl = *expr->decls_begin();
     OptionalFileEntryRef first_decl_file_entry = GetFileEntry(first_decl);
+    OptionalFileEntryRef current_file_entry = GetFileEntry(CurrentLoc());
```

Comment:
It's a little tempting to capture the information and pass it to `iwyu_output` to do the full analysis, e.g. something like `ReportOverloadUse`. That way iwyu.cc can focus more on traversing the AST and collecting interesting nodes, and less on the policy for those nodes (i.e. if they're directly included or not).

But there's enough tangling with current file entry and current AST node that that will probably be tricky. So just throwing it out there as a presumably-bad idea :)

Reply:
Yeah, that makes sense. I think for me, looking at whether the decl is directly included is already a non-local concern. But I don't think we should do anything right now, the code is fine where it is, but I wish I had a clearer intuition for what kind of functionality should go where. It would be nice if there were clearer passes through IWYU's logic.